Issue: #6311 new check: OrderedProperties

thomassenger opened the pull request:
#6311 

The OrderedProperties Check is almost ready. 
There are the following open issues:
- builds only with: mvn package -Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true -DskipTests
-- I need help with failure: XdocsPagesTest.testAllCheckSections:356->validateCheckSection:451->validateParentSection:1300 config_misc.xml section 'OrderedProperties' should have matching parent expected:<[TreeWal]ker> but was:<[Chec]ker>

rnveach commented:
@thomassenger Please work on making CI green. We usually won't review the code until CI is green unless you are having issues.
As you work on code, please keep number of commits as 1.

rnveach commented:
> I need help with failure: XdocsPagesTest.testAllCheckSections:356->validateCheckSection:451->validateParentSection:1300 config_misc.xml section 'OrderedProperties' should have matching parent expected:<[TreeWal]ker> but was:<[Chec]ker>

Add your class to https://github.com/checkstyle/checkstyle/blob/f7ff7520eb861313d6365dff00cd791938bcf7a5/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L97 .

thomassenger commented:
I have a problem to get the code working. 
I would like to use function level synchronization. But PMD forces me to use block level. Than I break the Properties/Hashtable API, because here function level synchronization is used.

I am also forced by surefire to implement equals and hashCode. I would like to use the implementation from Properties/Hashtable. I do not have my own identity. They are also never called if I run my code.
How do I get around this problem?
Regards.

rnveach commented:
@thomassenger See UniquePropertiesCheck's internal class. Why do you need synchronize when our other check doesn't?

thomassenger commented:
OK , you do not need synchronize. It was just a warning.

thomassenger commented:
So I made the code "build". Especially I do not like the equals and the hashCode method.
So as requested now we have a basis to discuss.
Regards,
Thomas

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,189 @@
+            properties.load(inputStream);
+        }
+        catch (IOException ex) {
+            log(1, MSG_IO_EXCEPTION_KEY, file.getPath(),
```
I know other Checks catch and log a violation with this, but we should just wrap the exception in a CheckstyleException and re-throw it.
Just make sure re-throwing it still has the file name in the stack trace when `Checker` catches it.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,189 @@
+
+            final String propKey = (String) keys.nextElement();
+
+            //if (previousProp.compareTo(propKey) > 0) {
```
Please remove commented code.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,189 @@
+        for (int index = 0; index < fileText.size(); index++) {
+            final String line = fileText.get(index);
+            matcher.reset(line);
+            if (matcher.matches()) {
```
How reliable is this at finding the line number?
What if my key was `a` and I had a bunch of values before `a=` that had `a` in it? Is it possible it won't pick the right line?
Can keys be multi-line like values?
If the same key is used twice, it will report the wrong line.
If this isn't reliable, we shouldn't use it as people will just complain of false line numbers.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,189 @@
+ * This implementation checks that properties files are sorted.
+ * It is using the binary order from String.compare()
+ *
+ * @author Thomas Senger
```
No author tags.
Also add your check to https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java#L70 and align it with the xdoc you have written.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,189 @@
+        }
+
+        @Override
+        public boolean equals(Object o) {
```
Why are equals and hashcode needed? If there is a violation we can just suppress it like `UniquePropertiesCheck` does.

thomassenger replied:
No, I do not need equals and hashCode. I was not able to suppress it. I have got something, that suppression is not allowed...
I struggling with the strict rules. It is my first contribution. Please give me some advice here.

rnveach replied:
Which CI gave you the violation that you needed equals/hashcode? Suppression is dependent on who is requiring it.

thomassenger replied:
it was spotbugs, EQ_DOESNT_OVERRIDE_EQUALS
I was able to solve it by editing spotbugs-exclude.xml

rnveach commented on src/main/resources/com/puppycrawl/tools/checkstyle/checks/messages.properties:
```diff
@@ -7,6 +7,7 @@ final.parameter=Parameter {0} should be final.
 forbid.escaped.unicode.char=Unicode escape(s) usage should be avoided.
 noNewlineAtEOF=File does not end with a newline.
 properties.duplicate.property=Duplicated property ''{0}'' ({1} occurrence(s)).
+properties.notSorted.property=Property key: ''{0}'' is not in right order with previous property ''{1}''.
```
I don't think the `:` is needed. Violation should be a complete sentence and not 2 separate thoughts.
`not in right order with previous property` => `not in the right order with the previous property`

thomassenger replied:
I removed the : and corrected the missing "the"
The key name 0 and 1 i kept. I always search for it.
Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,300 @@
+/**
+ * This implementation checks that properties files are sorted.
+ * It is using the binary order from String.compare()
+ */
```
Tests don't need javadocs.
I would remove them all except for any reflection that needs to be done.

thomassenger replied:
Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,300 @@
+     * valueOf() is uncovered.
+     */
+    @Test
+    public void testLineSeparatorOptionValueOf() {
```
This test shouldn't be needed since it outside the scope of your class.

thomassenger replied:
Deleted.
Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,300 @@
+     * method return value.
+     */
+    @Test
+    public void testNotFoundKey() throws Exception {
```
Why can't this use an input file instead of hacking the check?

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    @Test
+    public void testWrongKeyTypeInProperties() throws Exception {
```
Why can't this use an input file instead of hacking the check?

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    @Test
+    public void testEquals() throws Exception {
```
I assume this can go away if the equals/hashcode method is gone too.

rnveach commented on src/xdocs/checks.xml:
```diff
@@ -795,6 +795,9 @@
       <tr>
         <td><a href="config_misc.html#UniqueProperties">UniqueProperties</a></td>
         <td>Detects duplicated keys in properties files.</td>
+      </tr>      <tr>
```
Split this into 2 lines.

rnveach commented on src/xdocs/config_misc.xml:
```diff
@@ -1279,6 +1279,86 @@ void foo(String aFooString,
+          </table>
+      </subsection>
+
+      <subsection name="Examples" id="OrderedProperties_Examples">
```
Please expand this section with examples of good order and violations.

rnveach replied:
I don't see any changes yet, but I would start off easy before going into difficult examples. You are trying to introduce users to your check and how violation works.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,189 @@
+import com.puppycrawl.tools.checkstyle.api.FileText;
+
+/**
+ * This implementation checks that properties files are sorted.
```
`sorted` should be expanded upon.
Descending order is still sorted, etc..

thomassenger replied:
JavaDoc updated.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,189 @@
+            if (String.CASE_INSENSITIVE_ORDER.compare(previousProp, propKey) > 0) {
+
+                final int lineNumber = getLineNumber(fileText, propKey);
+                // Number of occurrences is number of duplications + 1
```
This check doesn't care about number of occurrences, right? So this comment seems pointless.

rnveach commented on pom.xml:
```diff
@@ -1804,6 +1804,7 @@
                 <param>com.puppycrawl.tools.checkstyle.checks.TranslationCheck*</param>
                 <param>com.puppycrawl.tools.checkstyle.checks.UncommentedMainCheck*</param>
                 <param>com.puppycrawl.tools.checkstyle.checks.UniquePropertiesCheck*</param>
+                <param>com.puppycrawl.tools.checkstyle.checks.SortedPropertiesCheck*</param>
```
You don't see something wrong with this line?

rnveach commented:
@thomassenger Please don't mark my points as resolved. Leave that to me to do to reverify everything you have done.
To points you have done, just reply "Done".

rnveach commented on pom.xml:
```diff
@@ -1821,6 +1822,7 @@
                 <param>com.puppycrawl.tools.checkstyle.checks.TranslationCheckTest</param>
                 <param>com.puppycrawl.tools.checkstyle.checks.UncommentedMainCheckTest</param>
                 <param>com.puppycrawl.tools.checkstyle.checks.UniquePropertiesCheckTest</param>
+                <param>com.puppycrawl.tools.checkstyle.checks.SortedPropertiesCheckTest</param>
```
You don't see something wrong with this line too?

thomassenger commented:
== https://travis-ci.org/checkstyle/checkstyle/jobs/478977402
It reports the synchronized problem.
So I was putting the synchronized back to keys and put.
I also added an in config/pmd.xml an exclude.

1. ERROR in /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java (at line 187)
	public Enumeration<Object> keys() {
	                           ^^^^^^
The method OrderedPropertiesCheck.SequencedProperties.keys() is overriding a synchronized method without being synchronized
----------
2. ERROR in /home/travis/build/checkstyle/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java (at line 196)
	public Object put(Object key, Object value) {

thomassenger commented:
Thank you for your work.
Next week I am busy and I have not time to continue.
Line numbers need more investigation from my side. I will try to solve it later.
Regards,
Thomas

romani commented:
Please reply as done to each raised items to let know that you are ready for review again.
CI should be green, please work to satisfy automatic review first.

rnveach commented:
When CI is green I will re-review unless you need help. You may want to squash all commits to 1 and rebase off the latest master as there are conflicts.

rnveach commented:
@thomassenger Please squash the commits into 1, don't leave a huge pile of commits in the PR. The branch is still unmergable and CI is still failing.

thomassenger commented:
Hello, I squash all commits into one. I cannot see your last requested change anymore.
The last change I did was the line numbering.
The build is still in progress. Lets see what the outcome is. Locally was the build fine.
Regards,
Thomas

rnveach commented:
> I cannot see your last requested change anymore.

Click the "Show resolved" and "Load more..." to show the requested changes that are now being hidden.

thomassenger commented:
Hello, I am trying to understand the failed ci/circleci. The test TranslationCheckTest.testTranslationFilesAreMissing runs locally fine. I do not understand the reported difference in list.

continuous-integration/travis-ci/pr: It reports that the issue number should refer to an PR. As far as I can see, is the issue number correctly at the commit message.
Many Thanks,
Thomas

rnveach commented:
There are no failures in circleci. Only IDEA, pitest3, and travis.

Travis failure is because issue number in commit message is wrong. You referenced 6312 and not 6311.

rnveach commented:
@thomassenger Do you still need help with CI?

thomassenger commented:
Hello,
Thanks for you question. I was on holiday and I also need help. 
I cannot figure out was the problem is on :
OrderedPropertiesCheck.java  if (!(keyList.contains(key))) 
OrderedPropertiesCheck.java  setFileExtensions(&#34;properties&#34;);

I know it is a test that mutating the code. I will now try to setup it locally.
For me the code looks fine.

Also you requested a change. I cannot see it anymore because I squashed the commits. You may need to request it again.

Thomas

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,202 @@
+         */
+        @Override
+        public synchronized Object put(Object key, Object value) {
+            if (!(keyList.contains(key))) {
```
`if (!(keyList.contains(key))) {` => `if (!keyList.contains(key)) {`

thomassenger replied:
I found by studying the PIT problems, that the keyList does not have to be unique. The order is important. So I removed the if complete.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,221 @@
+    public void testBinaryOrder() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(OrderedPropertiesCheck.class);
+        final String[] expected = {
+            "5: " + getCheckMessage(MSG_KEY, "key", "\u00C3\u00A4"),
```
`"\u00C3\u00A4"` => `"Ã¤"`

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,202 @@
+
+        String previousProp = "";
+
+        final Enumeration<Object> keys = properties.keys();
```
`properties.keys();` => `properties.keySet().iterator()`

thomassenger replied:
If I would do this, than the check will not work. Because keySet() creates a new Set by using new Hashtable.KeySet(). On a Hashtable is the order not guaranteed.

rnveach replied:
Then you will have to suppress the inspection.

thomassenger replied:
suppressed with
@noinspection EnumerationCanBeIteration

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,202 @@
+         * @param value the value
+         * @return the previous value of the specified key in this hashtable,
+         *      or null if it did not have one
+         * @Throws NullPointerException - if the key or value is null
```
lowercase `@throws`

thomassenger replied:
done

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,202 @@
+        /**
+         * Holding the keys in the same order than in the file.
+         */
+        private final List<Object> keyList = new ArrayList<>();
```
> 173: keyList Non-serializable field 'keyList' in a Serializable class 

This inspection may have to be suppressed.

thomassenger replied:
I was setting it to transient. It does not need to be serialized.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,202 @@
+     * Construct the check with default values.
+     */
+    public OrderedPropertiesCheck() {
+        setFileExtensions("properties");
```
> `+OrderedPropertiesCheck.java.html:<td class='covered'><pre><span  class='survived'>        setFileExtensions(&#34;properties&#34;);</span></pre></td></tr>`

This one is easy. Add a test where the file given has a `.java` extension and the file contains very basic java class. It should produce no violations/exceptions.

thomassenger replied:
I added an test that checks if only ".properties" is used.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,202 @@
+         */
+        @Override
+        public synchronized Object put(Object key, Object value) {
+            if (!(keyList.contains(key))) {
```
> `+OrderedPropertiesCheck.java.html:<td class='covered'><pre><span  class='survived'>            if (!(keyList.contains(key))) {</span></pre></td></tr>`

One of your tests must have a property file with a key that appears twice in the file. Since you remove duplicate keys, you must show that the 2nd instance of the key won't produce a violation if it is also out of order like the 1st instance.

thomassenger replied:
I added in a property file an double key. I removed this if, because it works also if the list is non unique.

romani commented:
@thomassenger , CI is green, I hope you are not exhausted, to pass review by humans

thomassenger commented:
Yes I see. Unbelievable. ;-)

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,210 @@
+        final Pattern keyPattern = getKeyPattern(keyName);
+        int humanIndex = 0;
+
+        for (int index = 0; index < fileText.size(); index++) {
```
Question was never answered at https://github.com/checkstyle/checkstyle/pull/6343#discussion_r245650215.
> How reliable is this at finding the line number?
What if my key was `a` and I had a bunch of values before `a=` that had `a` in it? Is it possible it won't pick the right line?
Can keys be multi-line like values?
If the same key is used twice, it will report the wrong line.
If this isn't reliable, we shouldn't use it as people will just complain of false line numbers.

thomassenger replied:
I was looking into  UniquePropertiesCheck. I found also 2 other implementation having the same pattern.
This implementation relies on the FileText. It loops and count what this implementation returns. My implementation will stop if the first match was found. (no multi key)
I can try to implement one more test to prove your concern above. I will try to verify the getKeyPattern() method. Because this seems unclear here.

thomassenger replied:
I implemented for getLineNumber two new validations.
1) check previous property key followed by property key
2) the starting line number is the line number of the last found problem.
In testfile InputOrderedProperties2EmptyValue.properties I inserted repeating keys to verify the implementation.
This should help to find the right line number for reporting a problem.

rnveach replied:
What about this concern?
> Can keys be multi-line like values?

thomassenger replied:
Today I worked on multi-line key values.
Java Property class can handle this. The key string contains a real newline. The FileText returns an already quoted string `\\n`. Therefore the regexp will not find a match. I solved this by quoting the keys. Than also the reporting log method will show the message correct.  This looks not good and you may have an better idea.

thomassenger replied:
Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    @Test
+    public void testNotFoundKey() throws Exception {
```
Question wasn't responded to at https://github.com/checkstyle/checkstyle/pull/6343#discussion_r245652053
> Why can't this use an input file instead of hacking the check?

thomassenger replied:
OK, I will rework this.

thomassenger replied:
Test deleted. Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    @Test
+    public void testWrongKeyTypeInProperties() throws Exception {
```
Question wasn't responded at https://github.com/checkstyle/checkstyle/pull/6343#discussion_r245652527
> Why can't this use an input file instead of hacking the check?

thomassenger replied:
I think I did some complicated stuff here to get the test coverage green. I think it is not required any more, because I removed the if from the put method.
I will rework this.

thomassenger replied:
Test deleted. Done.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,210 @@
+
+            final String propKey = (String) keys.nextElement();
+
+            //if (previousProp.compareTo(propKey) > 0) {
```
This item has not been done from https://github.com/checkstyle/checkstyle/pull/6343#discussion_r245649833 .
> Please remove commented code.

thomassenger replied:
OK, I will. My plan was to allow case sensitive an case insensitive check. But I think I will postpone it, because this MR getting too long to fulfil.

thomassenger replied:
deleted. Done

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/orderedproperties/InputOrderedProperties.txt:
```diff
@@ -0,0 +1,3 @@
+# txt files will not processed. This content is not relevant.
+1=val
+1=foo
```
This is not a good test case since this is still a valid properties file on the inside.
Either make this an invalid properties file or make it out of order so if it was processed there would be a violation.

thomassenger replied:
This file is used for testShouldNotProcessFilesWithWrongFileExtension(). The content of this file is not relevant. I could remove line 2 and 3?

rnveach replied:
The contents do matter. Imagine file extension check failed and this file was loaded as a property file. It would be a valid file, be loaded, and processed, and not produce any violations. Your check doesn't print violations on duplicate keys, just unsorted. This is why I said this is not a good input file.
Removing all but line still produces the same result. This is why I said to either make thsi an invalid properties file so it can't be loaded, or make it out of order to produce a violation if the file extension check fails.

thomassenger replied:
The file is unsorted, to product an possible violation.

thomassenger replied:
Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,228 @@
+        final Constructor<?> constructor = orderedPropertiesClass.getDeclaredConstructor();
+        constructor.setAccessible(true);
+
+        final Object orderedProperties = constructor.newInstance();
```
Constructor isn't private so why is this needed?

thomassenger replied:
Because of "I do not know". There are 12 other tests doing it in the same way. I was copying it from somewhere.
I will change it.

rnveach replied:
Those tests are probably dealing with a private constructor. We are working towards any reflection needs to be documented why it is needed. It is better we try to work without it. We just recently started removing powermock as it is causing issues with other tests and pitest. See https://github.com/checkstyle/checkstyle/issues/6439

thomassenger replied:
Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    @Test
+    public void testLineNumbers() throws Exception {
```
Why can't this use an input file instead of hacking the check?

thomassenger replied:
Hmmm lets see If I find a better solution. testLineNumbers() is private and my goal was to test the return of this method only. May be that I will keep it like this.

thomassenger replied:
Done.

rnveach replied:
@thomassenger This isn't done because the check is still using some reflection and it hasn't been stated why you can't check the method through inputs instead of using reflection.

thomassenger replied:
It was difficult to remove. Now I found out that there is a useful report on PIT. Using it I was able to understand better what PIT is doing. The new test is testKeepForLoopIntact(). It works without reflection and trick PIT.

thomassenger replied:
Done.

rnveach commented on .ci/pitest.sh:
```diff
@@ -37,7 +37,9 @@ pitest-annotation|pitest-design \
   mvn -e -P$1 clean test org.pitest:pitest-maven:mutationCoverage;
-  declare -a ignoredItems=();
+  declare -a ignoredItems=(
+    "OrderedPropertiesCheck.java.html:<td class='covered'><pre><span  class='survived'>            if (!prevKeyFound &#38;&#38; prevKeyPattern.matcher(line).matches()) {</span></pre></td></tr>"
```
You can't place this in this area as this is for pitest that produce no results.
You must move your specific pitest out like the others below this.

rnveach commented on config/suppressions.xml:
```diff
@@ -98,5 +98,8 @@
     <suppress checks="ExecutableStatementCount" files="HandlerFactory\.java"/>
 
+    <!-- can't split long messages between lines -->
+    <suppress checks="RegexpSinglelineJava" files="OrderedPropertiesCheckTest\.java"/>
```
What lines can you not split? There should be no way lines can't be split in java. We don't have this issue in our other tests, even those with long strings.

thomassenger replied:
You are right, I changed my test. It is not required any more. I removed it.

thomassenger replied:
Done.

thomassenger commented:
Hello rnveach, 
Could you please help me, with : ./.ci/travis/travis.sh verify-no-exception-configs
The xmlstarlet cannot download dtd's from the xml file.
I have no idea why. Firefox can download it. It has also nothing to do with my  OrderedPropertiesCheck.
Regards, Thomas

rnveach commented:
> verify-no-exception-configs

Do not worry about this CI item until code is ready to be merged. It just means this new module has to be appear in another config outside this repo.

> The xmlstarlet cannot download dtd's from the xml file.

This is not an issue. The difference below in the CI is the reason for the failure because your check isn't listed in this external config.

romani commented:
Travis failure is indeed confusing, can we upgrade script like proposed at https://stackoverflow.com/a/45310101/1015848 ? to avoid such warnings ?

It would be better to print some message that explains why build is failed, like "Change i noticed, in config file all Checks must be present"

@thomassenger , you can create PR to checks-nonjavadoc-error.xml to add your new Check. We will merge it after we merge this PR.

romani commented:
@thomassenger , if no activity till end of month, we will close this PR just to keep PRs list managable.
it is completely ok to reopen PR ones you have time to continue.

thomassenger commented:
Hello Romani,
thanks for your reminder. I would really like to get ready with this check. My project just finished last week.
I have really a problem to get around PMD warnings or hard to understand mutation tests. I do not really know what kind of PR to the checks-nonjavadoc-error.xml I should create? 

I did two things:
1) I removed some tests not to use reflections. (
orderedPropertiesClass.getDeclaredConstructor(); 
constructor.setAccessible(true);)

2) I reworked the getLineNumber to make it more robust findinf the right line. To prove this I extended ‘InputOrderedProperties.properties’ having repeating keys.
I also changed testLineNumbers() not use reflections creating an instance. It uses only reflection to make the method directly testable. This is required to see directly the impact of this implementation and I go an other failure of an mutation test.

I home we can close this PR soon. I may ask to organize an telko to be able to work together and solve build problems quick. I have time next week 23.04. - 26.04.  UTC+2 (Germany)

Regards, Thomas

PS: I have no access to wercker.com and travis-ci.org failed because of  checks-nonjavadoc-error.xml.

rnveach commented:
wercker has no failures, so you are good there. Wercker access is an issue with them as they don't allow to give out guest access.
If you have any issues with CI just let us know and we can help.

rnveach commented on .ci/pitest.sh:
```diff
@@ -36,12 +36,20 @@ pitest-annotation|pitest-design \
+pitest-misc)
+  mvn -e -P$1 clean test org.pitest:pitest-maven:mutationCoverage;
+  declare -a ignoredItems=(
+  "OrderedPropertiesCheck.java.html:<td class='covered'><pre><span  class='survived'>            if (!prevKeyFound &#38;&#38; prevKeyPattern.matcher(line).matches()) {</span></pre></td></tr>"
```
We prefer no surviving mutants.
I will look into this and try to help.

rnveach replied:
The code can be changed to `if (true && prevKeyPattern.matcher(line).matches()) {` and no tests are failing. So you need a test to prove that the code can't be changed.

Unless the code changes, the easiest way to remove this mutation is to make the code like:
`````
if (prevKeyFound) {
...
} else {
...
}
`````

thomassenger replied:
I will look into it.
The `if(true &&` trick does not look good and will not solve the problem.
Because 1) 
`[checkstyle] [ERROR] checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:152:31: Expression can be simplified. [SimplifyBooleanExpression]`

and 2) the mutation problem persists.

At the `if (prevKeyFound)` suggestion I am still working. The test coverage is not 100% and mutation score is 99 -> not 100. So here the (local) build failes.

rnveach replied:
@thomassenger If you misread what I said. The `if(true` is what pitest is doing behind the scenes, not what I wanted you to change the code to.
The code can be changed to `if (prevKeyFound) { } else { }` to remove the mutation.

thomassenger replied:
I changed the code to `if (prevKeyFound) { } else { }` now we will not have timed out mutations. But PIT will try to count backward through the file and change all if to else.
The code looks not very good. I do not know if it is a good strategy to seek for the previous and current key.
The testLineNumbers() is also not removed yet.

thomassenger replied:
I found a different solution.
Done.

rnveach commented on src/xdocs/config_misc.xml:
```diff
@@ -1320,6 +1320,111 @@ void foo(String aFooString,
 
+    <section name="OrderedProperties">
+      <subsection name="Description" id="OrderedProperties_Description">
+          <p>Since Checkstyle 5.8</p>
```
We are working towards 8.20, you should probably put 8.21 as I don't think this will done before the release next Saturday.

thomassenger replied:
Done.

rnveach commented on src/xdocs/config_misc.xml:
```diff
@@ -1320,6 +1320,111 @@ void foo(String aFooString,
+                  <td>file type extension of the files to check.</td>
+                  <td><a href="property_types.html#stringSet">String Set</a></td>
+                  <td>.properties</td>
+                  <td>5.7</td>
```
Same.

thomassenger replied:
Done.

rnveach commented on config/pmd.xml:
```diff
@@ -379,11 +379,12 @@
-                value="//ClassOrInterfaceDeclaration[@Image='UniqueProperties']
-                           //MethodDeclarator[@Image='put']"/>
+        value="//ClassOrInterfaceDeclaration[@Image='UniqueProperties'
+        or @Image='SequencedProperties']//MethodDeclarator[@Image='keys' or @Image='put']"/>
```
Please indent this last line by 2 spaces.

thomassenger replied:
Done.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,222 @@
+                prevKeyFound = true;
+            }
+
+            if (prevKeyFound && keyPattern.matcher(line).matches()) {
```
if `prevKeyFound` is set and `keyPattern.matcher` fails, shouldn't `prevKeyFound` also be reset?

rnveach replied:
If it is reset, you would have to recheck `prevKeyPattern` since you could have something like:
````
b=1
b=2
a=0
````

thomassenger replied:
I added a test for this. prevKeyFound does not need to reset, because of the break.

thomassenger replied:
Done.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,222 @@
+     * @param file the file to be processed
+     * @param fileText the contents of the file.
+     * @throws CheckstyleException
+     * I like to keep the order - keySet is not possible
```
This doesn't seem to be the reasoning for a checkstyle exception. Please update.

thomassenger replied:
Done.

rnveach replied:
@thomassenger
> I like to keep the order - keySet is not possible

This shouldn't be connected to the `@param` either. If there is no next `@` the text goes with the previous when it is printed to the javadoc.
Either put it after the full `@noinspection` or remove it.

rnveach commented on src/xdocs/checks.xml:
```diff
@@ -597,6 +597,10 @@
       </tr>
+      <tr>
+        <td><a href="config_misc.html#OrderedProperties">OrderedProperties</a></td>
+        <td>Detects if keys in properties files are in correct order.</td>
```
Please synchronize all your descriptions of the check to be the same. (Java, xdoc, checks)

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/orderedproperties/ExpectedOrderedProperties1OrderKey.properties:
```diff
@@ -0,0 +1,3 @@
+key1=value2
```
If these files are used somewhere please point me to them. We should only need `Input` files here.

thomassenger replied:
I removed this file.
Done.

rnveach commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/orderedproperties/ExpectedOrderedProperties4Numbers.properties:
```diff
@@ -0,0 +1,7 @@
+01=01
```
Same.

thomassenger replied:
I removed this file.
Done.

thomassenger commented:
I do not like the multi-line requirement and its solution.
Handling of newline in keys is not ideal. In
FileText.java Line 132 reader.readline() read it as characters (not as \n)

Also AbstractModuleTestSupport Line 238
variable `stream` having correct \n
variable `actuals` does not have correct \n

output on command line is correct.

Regards,
Thomas

romani commented:
I have never seen multiline properties, I even was not aware  that it is possible.
I would rather make it as limitation of Check in first version. If some user come up with request to support it, we can make it, such user will explain reason multiline key and we will put this in documentation to other users to be aware. 

@rnveach , please share your thoughts on this. I do not see where you demand it be supported by Check. I think it is miscommunication. You really mean support if easy and if not easy declare it in documentation as temporal limitation.

rnveach commented:
It was more a question/concern then a demand.  https://github.com/checkstyle/checkstyle/pull/6343#discussion_r277176164
I assume if it wasn't possible then java's parser wouldn't have parsed it and loaded it. I was more looking for flaws in how he identified the line the violation should be at.
I am fine without it.

romani commented:
@thomassenger , you are ok to not support muliline keys, BUT please declare it in documentation as limitation to keep implementation simple as it is not clear a reason to use multine keys in practice.

rnveach commented:
@romani Feel free to start reviewing. Regression has not been done yet.

romani commented on .ci/jsoref-spellchecker/whitelist.words:
```diff
@@ -80,6 +80,7 @@ ation
 atlassian
 attlist
 attr
+auml
```
please remove from whitelist.words all extra words (only `orderedproperties` is ok to stay), please use good words in testing

thomassenger replied:
Done.

thomassenger commented:
Hello Romani,
Thank you. Yes I agree. I will remove the newline functionality and add this as a limitation to the documentation.
Regards,
Thomas

thomassenger commented:
I think all changes are done. I so no action on my side.
Regards,
Thomas

romani commented on src/xdocs/config_misc.xml:
```diff
@@ -1320,6 +1320,118 @@ void foo(String aFooString,
 
+    <section name="OrderedProperties">
+      <subsection name="Description" id="OrderedProperties_Description">
+          <p>Since Checkstyle 8.21</p>
```
please use indentation as 2.

thomassenger replied:
done

romani commented on src/xdocs/config_misc.xml:
```diff
@@ -1320,6 +1320,118 @@ void foo(String aFooString,
+              a=97
+              key=107 than nothing
+              key.sub=k is 107 and dot is 46
+          </source>
```
please extend example to show case with violation as `key.p=value //violation`

thomassenger replied:
done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,227 @@
+ *       </li>
+ *   </ul>
+ * Known limitation: The key should not contain a newline.
+ * The string compare will work, but not the line number reporting.
```
please move it up, to be before example.

thomassenger replied:
done

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:
```diff
@@ -0,0 +1,184 @@
+     */
+    private static String getFileNotFoundDetail(File file) {
+        // Create exception to know detail message we should wait in
+        // LocalisedMessage
```
please make 2lines comment as single line, you have enough space

thomassenger replied:
done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,227 @@
+        }
+        catch (IOException ex) {
+            log(1, MSG_IO_EXCEPTION_KEY, file.getPath(),
+                    ex.getLocalizedMessage());
```
please move it to previous line, you have plenty of vertical space

thomassenger replied:
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,227 @@
+    /**
+     * Pattern matching single space.
+     */
+    private static final Pattern SPACE_PATTERN = Pattern.compile(" ");
```
please explain this pattern, such simple pattern usually do not regexp.

thomassenger replied:
This space pattern is used to construct a key pattern to search for keys in FileText. This pattern was taken from UniquePropertiesCheck and works well.

thomassenger commented:
the requested report 
groovy diff.groovy --localGitRepo ~/git/github/checkstyle/ --baseBranch upstream/master --patchBranch master --config my_check.xml --listOfProjects projects-to-test-on.properties

[reports.zip](https://github.com/checkstyle/checkstyle/files/3151112/reports.zip)

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,228 @@
+import com.puppycrawl.tools.checkstyle.api.FileText;
+
+/**
+ * This implementation checks that keys of properties files are sorted.
```
First sentence is very well used in javadoc. Please use `Detects if keys in properties files are in correct order.` from xdoc.

thomassenger replied:
I used your text.

romani commented on .ci/jsoref-spellchecker/whitelist.words:
```diff
@@ -1183,6 +1184,7 @@ stylesheets
 subdir
 subelements
 subext
+subkey
```
can we avoid this word usage ?

thomassenger replied:
I removed it.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:
```diff
@@ -0,0 +1,228 @@
+ *           The capital A is on 65 and the lowercase a is on position 97 on the ascii table.
+ *       </li>
+ *       <li>
+ *           Key and subkey are in correct order here, because only keys are relevant.
```
what is `subkey` ?
Is this is term from property files structure/specification ? if so please educate users  by link. If not please use exact name from example.

romani commented on src/xdocs/config_misc.xml:
```diff
@@ -1320,6 +1320,118 @@ void foo(String aFooString,
+      <subsection name="Description" id="OrderedProperties_Description">
+        <p>Since Checkstyle 8.21</p>
+        <p>
+          Checks if property keys (left from equal sign) are sorted.
```
please use the same summary as I proposed in for javadoc.

thomassenger replied:
I synced the docu.

romani commented:
@thomassenger , ping

thomassenger commented:
Hello Romany, I was busy last week and continue soon.
Regards

thomassenger commented:
To publish diff.groovy report seem to be difficult. The master branch should be orphaned. But I have already an master branch. 
The pages also does not exists. Can I publish is somewhere else?
The changed the docu as requested.

romani commented:
code is ok to merge.
Please squash commits in one, we are very close to PR merge.

We need regression report to make sure Check is not failing on real code.
provided link has no violations in report, and it is not clear on what project you did run.
Please provide details of your difficulties with our tool execution, we will help.

> thomassenger:master

yes, just make `git checkout -b issue-6311` on master it will create new branch on your local.
and do `git checkout master` and `git reset HEAD~2 --hard` (this will reset your master to state of our master)
run tool.

romani commented:
Wercker is failing:
`Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module MissingJavadocMethod - Unable to instantiate 'MissingJavadocMethod' class, it is also not possible to instantiate it as .......... Please recheck that class name is specified as canonical name or read how to configure short name usage https://checkstyle.org/config.html#Packages. Please also recheck that provided ClassLoader to Checker is configured correctly.
    at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild (TreeWalker.java:136)`

but our master build is not failing, please rebase your code to most recent our master code to fix issue.

Teamcity failure can be ignored, smth wrong on Teamcity side.

romani commented:
@thomassenger , we will have a release at weekend, it would be awesome to finish this PR.

thomassenger commented:
> We need regression report to make sure Check is not failing on real code.

I merged, squashed and created a new branch issue-6311 on my fork. 
This line above shows me that you believe that we use checkstyle including my new OrderedPropertiesCheck somewhere. This is not the case. I was running the diff report against: https://github.com/checkstyle/checkstyle
So the diff report will also show no violations. It looks similar to the one I published to  https://thomassenger.github.io/testrepo/.
I have no active project using checkstyle. Let me see what I can do.
Regards.

rnveach commented:
@thomassenger 
> I made a different repo: https://thomassenger.github.io/testrepo/

You seemed to have uploaded the raw reports generated by maven plugin and not the ones produced by the diff-report utility.
See http://rveach.no-ip.org/checkstyle/regression/reports/247/ as an example of the report we are expecting to be generated. The report generated should be in like a `/reports/diff` directory.

> This line above shows me that you believe that we use checkstyle including my new OrderedPropertiesCheck somewhere. This is not the case.
So the diff report will also show no violations. It looks similar to the one I published to 

You must modify the XML config file you give to `diff.groovy` to include your new check so it is run against different code bases and we can verify it behaves correctly with the violations it produces.
Looking at https://thomassenger.github.io/testrepo/reports/master/guava/checkstyle-checker.xml , the file has not been modified.

> I was running the diff report against: https://github.com/checkstyle/checkstyle

As [discussed in the readme](https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#diffgroovy-diff-report-generation) you must modify `projects-to-test-on.properties` to include all projects in the file, so you will be running regression against many projects then just the main checkstyle repo.

Sorry this process hasn't worked well for you. We welcome any additions to the readme to make this process more seamless.

thomassenger commented:
Thank you rnveach for your help. Now its is getting clear.
Now I am running the diff.groovy with more and more projects commented out (projects-to-test-on.properties)
Currently I add also some check rules to my_check.xml.
It is deployed to https://thomassenger.github.io/testrepo/ (I move it later)

rnveach commented:
> If I add my checker module name="OrderedProperties" to my_check.xml the groovy script will fail. I believe because of the missing check at the upstream/master branch.

You must run a patch-only difference report because the check doesn't exist in master for a full difference. The report example I gave is similar.
https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#diffgroovy-diff-report-generation
See the section that says `To generate the report only for the patch branch which contains your changes`. You need to use `--patchBranch` and `--patchConfig` and `--mode single`.

thomassenger commented:
I made a new diff report having --patchConfig and --mode single.
The report was pushed to github.com:thomassenger/checkstyle.git master branch
I open a new pull request on branch "issue-6311"
I think is is getting in the right direction.
Regards.

rnveach commented:
@thomassenger I don't know what happened but it looks like you pushed to this PR and overwrote your original work.
https://github.com/checkstyle/checkstyle/commit/b829c2471ac1d05e701a62e7f8a290db7d46ad2a was your original commit

rnveach commented:
It looks like this PR was moved to https://github.com/checkstyle/checkstyle/pull/6781 .
@thomassenger in the future, it is best to put the difference report in some place other than the normal repo you use to make PRs with. I will close this PR and we can use the fresh new one and mention this is where the report is located.

thomassenger commented:
> It looks like this PR was moved to #6781 .

Yes, this is what I understand.

> @thomassenger in the future, it is best to put the difference report in some place other than the normal repo you use to make PRs with. I will close this PR and we can use the fresh new one and mention this is where the report is located.

You confuse me.  That is why I used the testrepo in the first place.
To get everything right I moved all the code to a branch "issue-6311" and made a new PR to have the master branch free for the diff report.
The diff report is: https://thomassenger.github.io/checkstyle/

Anyhow. Lets get this done.
Regards.
Thomas

rnveach commented:
Thanks. It has been a while since I used github.io . I thought I used a different repo.
